### DISCUSSION TITLE
[Propousal] <> diamond operator

### DISCUSSION THREAD

**Discussion by redradist** [dotnet/csharplang#2935] [Language Ideas]
When we have generic class like this:

```csharp
List<string> cars = new List<string>();
```

Why should we repeat ourselves ?

Let's simplify it like Java did:

```csharp
List<string> cars = new List<>();
```

Generic type parameter can be deduced from the context ...

```csharp
IDictionary<string, int> carOwnerIds = new Dictionary<>();
```

Adding this feature would be according the principle **DRY** - _Do not Repeat Yourself_

**Comment by HaloFour** [comment]
Doesn't seem terribly necessary given C# has `var` and will be getting target-typed `new`:

```cs
var cars = new List<string>();
Dictionary<string, int> carOwnerIds = new();
```

If you want to see something fun, the following is legal in Java.  Guess what it does?

```java
var list = new ArrayList<>();
```

**Comment by HaloFour** [comment]
Here's a championed issue where they are aiming to explore this kind of issue: https://github.com/dotnet/csharplang/issues/1349

_Edit:_
Oops, that is specifically for generic methods, not constructors/types.  I think https://github.com/dotnet/csharplang/issues/281 proposes it for constructors.

**Comment by alrz** [comment]
```cs
var list = new ArrayList<>();
```
Tried it a while back and looks like it infers `ArrayList<Object>`

I think it would make sense to make that an error if it can't be inferred from arg list or the target-type.

**Comment by redradist** [comment]
> Doesn't seem terribly necessary given C# has `var` and will be getting target-typed `new`:
> 
> ```cs
> var cars = new List<string>(); // cars type inferenced as List<string>
> Dictionary<string, int> carOwnerIds = new(); // Just simplified construction carOwnerIds type
> ```

@HaloFour I agree that we have a lot of inference already, but consider the following case with inferencing to interface type:
```csharp
IEnumerable<string> cars= new List<>(); // Seems like this feature actually could be helpful
```

What do you think ?

**Comment by huoyaoyuan** [comment]
> inferencing to interface type:
> ```csharp
> IEnumerable<string> cars= new List<>(); // Seems like this feature actually could be helpful
> ```
> What do you think ?

In interface scenario, this feature is far not helpful, but impossible.
`string` is quite a special case because it's sealed.

In CLR, generics are strongly typed with covariance and contravariance. Given these:
```csharp
IEnumerable<BaseClass> = new YourSpecialList<???>();
IEnumerable<???> = new YourSpecialList<BaseClass>();
```
There are many types can fit them, and your special list can behavior differently by querying `typeof(T)`.

The feature works in Java just because **type erasure**. `C<A>` and `C<B>` are totally the **same** type in Java, but **different** in CLR. Even without variance, `class C : ISome<A>, ISome<B>` are totally valid in CLR.

The stronger the type system, there's more attention required with types. In JavaScript you don't even need to provide type when creating objects.

**Comment by redradist** [comment]
> In CLR, generics are strongly typed with covariance and contravariance. Given these:
> 
> ```cs
> IEnumerable<BaseClass> = new YourSpecialList<???>();
> IEnumerable<???> = new YourSpecialList<BaseClass>();
> ```

@huoyaoyuan It is pretty easy ... ;)

For example:
```csharp
IEnumerable<BaseClass> = new YourSpecialList<>();
```
, it will be deduced as BaseClass if possible, otherwise it will fail to deduce type and user has to explicitly specify Generic Parameter Type

For example:
```csharp
IEnumerable<> = new YourSpecialList<BaseClass>();  // compilation error
```
, it is not legal syntax !! Compilation error, type can be deduced only on right hand side of assigment expression

**Comment by Thaina** [comment]
> ```cs
> IEnumerable<string> cars= new List<>(); // Seems like this feature actually could be helpful
> ```
> 
> What do you think ?

```C#
var cars= new List<string>().AsEnumerable();
```

Yes it cannot be used in field declaration. But this would be an example of why we need #601

**Comment by huoyaoyuan** [comment]
> , it will be deduced as BaseClass if possible, otherwise it will fail to deduce type and user has to explicitly specify Generic Parameter Type

You may be meaning "try the generic parameter in the LHS, if it fits then use it". It's very fragile and limited.

**Comment by vladd** [comment]
@redradist 
> `IEnumerable<string> cars= new List<>();`

Well, it's not clear how the type inference is going to work. You cannot simply deduce that the missing type parameter is `string`! Imagine the following definitions:

```csharp
class Names<T> : IEnumerable<string> { ... }
```

or 

```csharp
class Sets<T> : IEnumerable<HashSet<T>> { ... }
```

The type inference is not allowed to just transfer the generic argument of the right side to the left side.

**Comment by alrz** [comment]
@vladd It's the same inference as follow: 
```cs
void M<T>(IEnumerable<T> s) {}

M(new List<string>()) // T is string
```
Only that it would be an upper-bound instead of lower-bound because the type flows inward.

So in those examples type inference could fail if the target type is not compatible with the source.

**Comment by vladd** [comment]
@alrz I respectfully disagree.

In the case you proposed, the logic is simple: the actual type is `List<string>`, and the list of interfaces it supports directly includes `IEnumerable<string>` which matches `IEnumerable<T>`.

But it's not the problem we have to solve! We have the backwards problem, which can be reduced to the following: given a type `X`, what should be the actual value of generic parameter (or parameter set) for generic class `C<P1, P2, ... Pn>` so that `C<P1, P2, ... Pn>` implements `X`? For example, given `X = IEnumerable<KeyValuePair<int, double>>`, and types

```csharp
interface IHelper<T> : IEnumerable<T> where T : class { }
class C<P1, P2, P3>: Dictionary<P1, P2>, Helper<P3> { ... }
```

is it really easy to deduce that P1 = int, P2 = double and P3 = any class?

**Comment by alrz** [comment]
@vladd I'm not trying to specify how type inference should work in any particular case, I'm saying that the same inference could be made from the target-type and the source, that doesn't imply any change or addition to the type inference spec itself. So if it's already an error, it will fail here as well.

**Comment by alrz** [comment]
This is similar to https://github.com/dotnet/csharplang/issues/2701 where the target-type is not an array, so that the source needs to be converted to the target.

**Comment by theunrepentantgeek** [comment]
@redradist, this is one of your motivating examples:

``` csharp
IEnumerable<string> cars= new List<>();
```

Where would we ever actually _want_ to write this code?

If we're declaring a local variable, this existing syntax is shorter and clearer:
``` csharp
var cars = new List<string>();
```

If we're passing the result _directly_ to a method, we've been able to do that with generics since C# 2.0:
``` csharp
void Foo<T>(IEnumerable<T> items) => ...

Foo(new List<string>());
```

If we're initializing a member field, those are almost always declared as the appropriate concrete type, and the new target-typed syntax will make that very simple:
``` csharp
private readonly List<string> cars = new();
```

So, as far as I can tell, the only relevant situation would be the rare case where the member field is being declared as a less specific type, where the current syntax is this:
``` csharp
private readonly IEnumerable<string> cars = new List<string>();
```

FWIW, I just did a troll through all the code I have available on my personal machine, and found exactly **_zero_** cases of a member field declared as `IEnumerable<anything>`.

_What am I missing here?_

**Comment by huoyaoyuan** [comment]
> What am I missing here?

```csharp
public IList<string> SomeProperty { get; } = new List<string>();
```
@theunrepentantgeek Quite a corner case, as all usages of the object can be done through interface. Only applies for limited scenarios.

**Comment by alrz** [comment]
I think for such expression there are at least three potential source of type inference,
```cs
var lazy = new Lazy<>(() => CreateFoo()); // argument list
var map = new Dictionary<,>() { { 1, 2 } }; // collection initializer
ICollection<C> c = new HashSet<>(); // target type
```
Removing the diamond syntax is explored at #281 but I think it would be hard to reason about.

Note that `new()` is too broad and might not work out with certain overloads, whereas `new C<>` could actually contribute to the overload resolution and therefore avoiding an ambiguous error,
```cs
void M(List<int> list) {}
void M(object o) {}

M(new() { }) // error: ambiguous call
M(new List<> { }) // ok: binds to the first method
```

**Comment by redradist** [comment]
> @redradist, this is one of your motivating examples:
> 
> ```cs
> IEnumerable<string> cars= new List<>();
> ```
> 
> Where would we ever actually _want_ to write this code?
> 
> If we're declaring a local variable, this existing syntax is shorter and clearer:
> 
> ```cs
> var cars = new List<string>();
> ```

Because in this case cars type is deduced as List<string>. but I want IList<string> or event IEnumerable<string>, but at the same time I do not want to repeat myself ...

This feature is incredibly simple ... ) Just take generic type parameter of left hand side type and try it for generic type parameter on right hand side ... That's all !!

And believe me it is very useful, I know, I have reach background in Java ;)

Also this feature is according DRY principle ;)

**Comment by HaloFour** [comment]
@redradist 

> Just take generic type parameter of left hand side type and try it for generic type parameter on right hand side ... That's all !!

It's simple in Java because the generic type arguments don't matter at all at the point of instantiation.  The compiler doesn't need to know that the class you are constructing requires the parameters in a certain order or constrained in a certain manner because those parameters don't exist at runtime and can be completely ignored.  In .NET this is very different because the generic type arguments are reified and both the compiler and the runtime care very much exactly what they are at the point of instantiation.  The compiler can't just copy&paste the generic type arguments from the LHS and try to apply them on the RHS.

> And believe me it is very useful, I know, I have reach background in Java ;)

Yes, but it was useful because it took Java an additional 7 years to bother adding local type inference.  I'd argue that it's much less relevant to any modern Java codebase.  I find myself rarely using it anymore, except for situations where type inference isn't available such as with fields.

**Comment by redradist** [comment]
> @HaloFour 
> The compiler can't just copy&paste the generic type arguments from the LHS and try to apply them on the RHS.

Why ? It is very simple !!
If this type does not fit or not satisfied constraints than produce error !!

I know that C$ Generic it is Run Time generics and I consider it is good, but on other hand I do not see why such simple technique as I described cannot be applied ?

**Comment by HaloFour** [comment]
@redradist 

> Why ? It is very simple !!

Is it?

```cs
public interface IFoo<T1, T2> { }

public class Bar<T1, T2> : IFoo<T2, T1> { }

public class Baz<T> : Bar<string, T> { }

static class Program {
    static void Main() {
        IFoo<int, string> foo = new Baz<>();
    }
}
```

The Java equivalent of that compiles just fine.

**Comment by redradist** [comment]
@HaloFour 
> 
> > Why ? It is very simple !!
> 
> Is it?
> 
> ```cs
> public interface IFoo<T1, T2> { }
> 
> public class Bar<T1, T2> : IFoo<T2, T1> { }
> 
> public class Baz<T> : Bar<string, T> { }
> 
> static class Program {
>     static void Main() {
>         IFoo<int, string> foo = new Baz<>();
>     }
> }
> ```
> 
> The Java equivalent of that compiles just fine.

Okay, and what is an issue ?
In Java it compiles and what ? ...

In C# we could make it either compilation error or make DAG and properly deduce the type at compile time ...

For example IFoo has two generic type parameters T1 and T2 than DAG would look like this:

```
T1(of IFoo) -> T2(of IFoo) -> string(of Baz)
T2(of IFoo) -> T1(of IFoo) -> T(of Baz)
```

Baz has only one generic parameter then obviously that we should try T2(of IFoo) -> T(of Baz) as string and if it fits than Baz deduced as Baz<string>

```csharp
public interface IFoo<T1, T2> { }

public class Bar<T1, T2> : IFoo<T2, T1> { }

public class Baz<T> : Bar<string, T> { }

static class Program {
    static void Main() {
        IFoo<int, string> foo = new Baz<>(); // Deduced as Baz<string>
    }
}
```

**Comment by momvart** [comment]
I agree with @redradist.
In many cases, for the sake of abstraction, we use interfaces of collections instead of concrete types. And it becomes really bothering when your class names are a bit large or generic types of generic types exist. We have declared the type before and again we have to write them for the concrete types and end up with huge expressions.

For example:
```csharp
IDictionary<MyKeyType, MyValueType.WithLargeNameNested<string>> dict;
dict = threadSafe 
           ? new ConcurrentDictionary<MyKeyType, MyValueType.WithLargeNameNested<string>>()
           : new Dictionary<MyKeyType, MyValueType.WithLargeNameNested<string>>();
```
It could be simply written as:
```csharp
IDictionary<MyKeyType, MyValueType.WithLargeNameNested<string>> dict;
dict = threadSafe ? new ConcurrentDictionary<>() : new Dictionary<>();
```

(Also, kotlin has a bit more complicated inference which its specs are defined [here](https://kotlinlang.org/spec/type-inference.html).)

**Comment by michael-hawker** [comment]
Yeah, now that we have type inference for `new` in C#9, it'd be great to have that work more for Generics as well. I don't even know if it needs an operator, if you call a generic method with known types to the compiler, you don't have to specify anything... So that'd leave us with:

```cs
IDictionary<string, object> = new Dictionary();
```

The generic arguments normally required for the `Dictionary` class should just grab them from the inferred interface type.

**Comment by michael-hawker** [comment]
Just found https://github.com/dotnet/csharplang/issues/1349 which seems related to this discussion.
